### PR TITLE
Modernize package infrastructure

### PR DESCRIPTION
**Pull request by PicoCentauri**
This PR updates a number of the infrastructure of the package:

1. Going from a tox multi-line string to an independent `tox.ini` file. The latter one is easier to maintain as a huge multi-line string and also consistent with the other projects.
2. Adding tox sections for _linting_, _building the documentation_ and _formatting the code_
3. Move `skmatter` to `src/skmatter`
4. Rename `docs/source` -> `docs/src` (same is in other lab-cosmo projects)
5. Translate from `setup.py`, `setup.cfg` into a single `pyproject.toml` file
6. Add a [documentation link action](https://github.com/readthedocs/actions/tree/v1/preview)
7. Add Python 3.11 to  the CI test matrix
8.  Consistent use of Python 3.10 for _docs_, _lint_ and _ReadTheDocs_ in the CI
9. Add tests for **windows** and **macos**
10. Update all gh-action dependencies to the lates version
11. Update `contributors.txt`
12. Update Rose's mail address
13. Use `isort` and `flake8` settings of the other lab-cosmo projects
14. Following 7 use 88 characters for all files including docs and comments
15. Remove `numpy` as an explict dependency. The whole package depends on `scikit-learn` which depends on `numpy`. I think we do not need to list `numpy` again in the dependencies if we stick with the same version as they do (which we should).

### CODE REVIEW

File: .github/workflows/lint.yml [hunk 1]
```diff
@@ -1,36 +1,20 @@
...
-        run: |
-          isort skmatter/*/*py -m 3 --tc --fgw --up -e -l 88 --check
+    - uses: actions/checkout@v3
+    - name: Set up Python
+      uses: actions/setup-python@v4
+      with:
+        python-version: "3.10"
+    - run: pip install tox
+
+    - name: Test Lint
```
Comment (Luthaf) [inline .github/workflows/lint.yml:19]:
```suggestion
    - name Lint the code
```

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -1,31 +1,35 @@
-name: Test
+name: Tests
 
 on:
   push:
     branches: [main]
   pull_request:
-    # Check all PR
+    branches: [main]
```
Comment (Luthaf) [inline .github/workflows/tests.yml:7]: I think we can still check all PR, even if not targeting main

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -1,31 +1,35 @@
...
     strategy:
       matrix:
-        python-version: ["3.8", "3.9", "3.10"]
+        os: ['ubuntu-latest']
+        python-version: ['3.8', '3.9', '3.10', '3.11']
+        include:
+          - os: macos-latest
+            python-version: '3.10'
+          - os: windows-latest
+            python-version: '3.10'
```
Comment (Luthaf) [inline .github/workflows/tests.yml:20]: Given that this is a pure Python package, the gain of checking code on different OS is much less than for packages with native code. I'm no against adding more builders, as long as they are fast and don't block CI

File: docs/src/conf.py [hunk 4]
```diff
@@ -45,15 +46,15 @@
...
+        nb_out = os.path.join(ROOT, "docs/src/read-only-examples", nb)
 
         # Skip any notebooks which already exist
         if not os.path.exists(nb_out):
             with open(nb_out, "w") as out_stream:
                 converted = exporter.from_filename(nb_in)[0]
                 out_stream.write(converted)
 
-import skmatter  # noqa
+import skmatter  # NoQa
```
Comment (Luthaf) [inline docs/src/conf.py:57]: This can be moved above, just below the `sys.path.insert`

File: docs/src/contributing.rst [hunk 5]
```diff
@@ -27,12 +27,9 @@ Running the tests
...
   cd <scikit-matter PATH>
-  # run unit tests
-  tox
-  # run the code formatter
-  black --check .
-  # run the linter
-  flake8
+  tox -e tests         # unit tests
+  tox -e lint          # code style
+  tox -e format        # format all files
```
Comment (Luthaf) [inline docs/src/contributing.rst:32]:
```suggestion
  tox  # run tests and linters
```

Comment (Luthaf) [reply docs/src/contributing.rst:32]: And I would document `tox -e format` separately

**Review by Luthaf — COMMENTED** [review]
The infrastructure changes looks good to me, I only skimmed over the documentation changes. 

`tox -e docs` and `tox -e format` should be documented somewhere (in Contributing maybe?)

**Comment by PicoCentauri** [comment]
Thanks for your review @Luthaf! I implemented the useful suggestions. If you are happy I would also like to have @rosecers opinion before we apply these changes.

File: .github/workflows/tests.yml (hunk 3 above)
Comment (Luthaf) [reply .github/workflows/tests.yml:20]: works for me

File: docs/Makefile
Comment (rosecers) [inline docs/Makefile:1]: Is this file no longer needed? Why?

File: docs/make.bat
Comment (rosecers) [inline docs/make.bat:1]: Is this file no longer needed? Why?

File: docs/src/installation.rst [hunk 6]
```diff
@@ -13,4 +13,5 @@ Install from source
   cd scikit-matter
   python setup.py install --user
 
-You're ready to import skmatter from your code! Have a look at the :ref:`api-reference` for how to use the code.
+You're ready to import skmatter from your code! Have a look at the :ref:`api-reference`
+for how to use the code.
```
Comment (rosecers) [inline docs/src/installation.rst:17]:
```suggestion
You're ready to import skmatter from your code! Have a look at the :ref:`api-reference` for how to use the code.
```

File: src/skmatter/_selection.py [hunk 7]
```diff
@@ -10,16 +10,9 @@
...
-    MetaEstimatorMixin,
-)
+from sklearn.base import BaseEstimator, MetaEstimatorMixin
 from sklearn.feature_selection._base import SelectorMixin
-from sklearn.utils import (
-    check_array,
-    check_random_state,
-    safe_mask,
-)
+from sklearn.utils import check_array, check_random_state, safe_mask
```
Comment (rosecers) [inline src/skmatter/_selection.py:15]: Fairly sure this will trigger the linters, correct?

File: src/skmatter/_selection.py [hunk 8]
```diff
@@ -57,19 +50,20 @@ class GreedySelector(SelectorMixin, MetaEstimatorMixin, BaseEstimator):
 
     score_threshold : float, default=None
         Threshold for the score. If `None` selection will continue until the
-        n_to_select is chosen. Otherwise will stop when the score falls below the threshold.
-        Stored in :py:attr:`self.score_threshold`.
+        n_to_select is chosen. Otherwise will stop when the score falls below the
+        threshold. Stored in :py:attr:`self.score_threshold`.
```
Comment (rosecers) [inline src/skmatter/_selection.py:54]:
```suggestion
        n_to_select is chosen. Otherwise will stop when the score falls below the  threshold. 
        Stored in :py:attr:`self.score_threshold`.
```

File: src/skmatter/_selection.py [hunk 9]
```diff
@@ -84,7 +78,8 @@ class GreedySelector(SelectorMixin, MetaEstimatorMixin, BaseEstimator):
     X_selected_ : ndarray,
                   Matrix containing the selected samples or features, for use in fitting
     y_selected_ : ndarray,
-                  In sample selection, the matrix containing the selected targets, for use in fitting
+                  In sample selection, the matrix containing the selected targets, for
+                  use in fitting
```
Comment (rosecers) [inline src/skmatter/_selection.py:82]: There are small, unnecessary formatting changes throughout here.

File: src/skmatter/_selection.py [hunk 10]
```diff
@@ -199,7 +194,8 @@ def fit(self, X, y=None, warm_start=False):
         if warm_start:
             if not hasattr(self, "n_selected_") or getattr(self, "n_selected_") == 0:
                 raise ValueError(
-                    "Cannot fit with warm_start=True without having been previously initialized"
+                    "Cannot fit with warm_start=True without having been previously"
+                    " initialized."
```
Comment (rosecers) [inline src/skmatter/_selection.py:198]: ?

File: src/skmatter/decomposition/_kernel_pcovr.py [hunk 11]
```diff
@@ -493,12 +480,14 @@ def score(self, X, Y):
...
-            \mathbf{K}_{VN} \mathbf{T}_N (\mathbf{T}_N^T \mathbf{T}_N)^{-1}  \mathbf{T}_V^T
+            \mathbf{K}_{VN} \mathbf{T}_N
+                (\mathbf{T}_N^T \mathbf{T}_N)^{-1} \mathbf{T}_V^T
             +\mathbf{T}_V(\mathbf{T}_N^T \mathbf{T}_N)^{-1}  \mathbf{T}_N^T
             \mathbf{K}_{NN} \mathbf{T}_N (\mathbf{T}_N^T \mathbf{T}_N)^{-1}
             \mathbf{T}_V^T\right]}{\operatorname{Tr}(\mathbf{K}_{VV})}
 
-        The negative loss is returned for easier use in sklearn pipelines, e.g., a grid search, where methods named 'score' are meant to be maximized.
+        The negative loss is returned for easier use in sklearn pipelines, e.g., a
+        grid search, where methods named 'score' are meant to be maximized.
```
Comment (rosecers) [inline src/skmatter/decomposition/_kernel_pcovr.py:490]: Will this render correctly?

File: setup.py
Comment (rosecers) [inline setup.py:1]: Why is this file no longer needed? I'm not seeing how removing it is beneficial.

**Review by rosecers — COMMENTED** [review]
Going through, all looks generally fine, but I'm worried that the new linting doesn't look as nice. Can we set it up with the same rules as before?

File: src/skmatter/_selection.py (hunk 9 above)
Comment (PicoCentauri) [reply src/skmatter/_selection.py:82]: Why unnecessary? There is a line width of 88 characters. This should be obeyed to please the linter.

**Comment by PicoCentauri** [comment]
> Going through, all looks generally fine, but I'm worried that the new linting doesn't look as nice. Can we set it up with the same rules as before?

Thanks. Black is exactly the same. The only thing which is slightly changed is how isort works. Packages are sorted according to: `FUTURE`, `STDLIB`, `THIRDPARTY`, `FIRSTPARTY`, `LOCALFOLDER`, which is the default. I think this order is much easier to read; especially the splitting between `THIRDPARTY` (`numpy`, `sklearn`) and `FIRSTPARTY` (`skmatter`).

File: .readthedocs.yaml [hunk 12]
```diff
@@ -5,17 +5,22 @@
...
 # Build documentation in the docs/ directory with Sphinx
 sphinx:
-  configuration: docs/source/conf.py
+  configuration: docs/src/conf.py
 
 # Optionally build your docs in additional formats such as PDF
 formats:
   - pdf
 
 # Optionally set the version of Python and requirements required to build your docs
```
[the remark this answers is not in the document]
Comment (PicoCentauri) [reply .readthedocs.yaml:22]: Yep.

File: MANIFEST.in [hunk 13]
```diff
@@ -1,2 +1,4 @@
-recursive-include skmatter/datasets/data/ *
-recursive-include skmatter/datasets/descr/ *
+recursive-include src/skmatter/datasets/data/ *
+recursive-include src/skmatter/datasets/descr/ *
+
+prune tests
```
[the remark this answers is not in the document]
Comment (PicoCentauri) [reply MANIFEST.in:4]: It does not include the test files in shipped package via PYPI or conda. We can remove this line and include the tests again... But, if people run the tests they will use the github repo and not the one from a package manager.

File: docs/Makefile
Comment (PicoCentauri) [reply docs/Makefile:1]: `tox -e docs` will build the docs in an isolated environment and takes care of the OS specific stuff. So no, we do not need this anymore.

File: docs/make.bat
Comment (PicoCentauri) [reply docs/make.bat:1]: Same as above.

File: docs/src/installation.rst (hunk 6 above)
Comment (PicoCentauri) [reply docs/src/installation.rst:17]: We can put this in one line but should not try to use 88 characters also in the docs? rst does not care about a new line.

File: src/skmatter/_selection.py (hunk 8 above)
Comment (PicoCentauri) [reply src/skmatter/_selection.py:54]: Same here 88 characters per line.

File: src/skmatter/_selection.py (hunk 10 above)
Comment (PicoCentauri) [reply src/skmatter/_selection.py:198]: 88 characters per line.

File: setup.py
Comment (PicoCentauri) [reply setup.py:1]: Our code is fully compatible with PEP517 and PEP660 🥳. We do not have to do any legacy builds which require a `setup.py` file: https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html


Also removing this file prevents user from doing `python setup.py install` which you should not do.

File: src/skmatter/decomposition/_kernel_pcovr.py (hunk 11 above)
Comment (PicoCentauri) [reply src/skmatter/decomposition/_kernel_pcovr.py:490]: [Looks good to me.](https://scikit-matter--172.org.readthedocs.build/en/172/pcovr.html#skmatter.decomposition.PCovR.score)

Screenshot 2023-03-28 at 23 00 58

**Comment by rosecers** [comment]
So I get the 88 char/line thing from a code-formatting standpoint, but as a code reader, I find it very frustrating. Industry-standard is 80 -- any arguments for/against doing that instead?

**Comment by ceriottm** [comment]
Ah, the good ole' Linus' rant.
FWIW I'm all for longer lines.

**Comment by rosecers** [comment]
I’d advocate for even longer, ~100char!

**Comment by PicoCentauri** [comment]
Okay, since this might be a longer and intensive discussion I suggest we keep the 88 chars/line in within this PR and open an issue about how we handle this in the future. One reason is that I would like to touch more files of the repo but this PR is already huge and complex.

**Review by rosecers — APPROVED** [review]
